src/libopensc/pkcs15.c

```diff
@@ -2574,6 +2574,31 @@ sc_pkcs15_get_supported_algo(struct sc_pkcs15_card *p15card, unsigned operation,
 	return info;
 }
 
+struct sc_supported_algo_info *
+sc_pkcs15_get_specific_supported_algo(struct sc_pkcs15_card *p15card, unsigned operation, unsigned mechanism, const struct sc_object_id *algo_oid)
+{
+	struct sc_context *ctx = p15card->card->ctx;
+	struct sc_supported_algo_info *info = NULL;
+	int ii;
+
+	if (algo_oid == NULL)
+		return NULL;
+
+	for (ii=0;ii<SC_MAX_SUPPORTED_ALGORITHMS && p15card->tokeninfo->supported_algos[ii].reference; ii++)
+		if ((p15card->tokeninfo->supported_algos[ii].operations & operation)
+				&& (p15card->tokeninfo->supported_algos[ii].mechanism == mechanism)
+				&& sc_compare_oid(algo_oid, &p15card->tokeninfo->supported_algos[ii].algo_id))
```

Comment:
you need to check for `1 == sc_compare_oid`

Reply:
Explicit check for == 1 added. 

Checking for boolean return values without == operator is normal in C language and not always considered as bad practice. I guess you want to make a clear difference with functions, which return 0 to inidicate success? (because CKR_OK and SC_SUCCESS are 0)

I put it other way around though: if (sc_compare_oid(...) == 1)
This style is mostly used in OpenSC and i personally like it better, I think the code is easier to read this way.

Reply:
Thanks!

Before this change, `sc_compare_oid` could have returned an error (negative), which would have been interpreted as success.

In general, using explicit checks is advisable in security related code... I haven't found the time yet to refactor OpenSC, however...

Reply:
Oh, and choosing `1 == x` over `x == 1` protects from overlooking the following typo `1 = x`, because it yields a compiler error...

Reply:
You are right, it could return SC_ERROR_INVALID_ARGUMENT so this check is required for it to work correctly, not only a style issue. I agree that it's good and more safe to use explicit checks. I'm more used to coding c++ where a bool function can only return true or false, not for example -1.

Which one is better, 1==x or x==1 could be a source of a neverending debate. I'd rather not go into it ;-)